Zero-fee commitments using v3 transactions (feature 40/41)

t-bast opened the pull request:
We introduce a new `channel_type` that leverages v3 (TRUC) transactions, pay-to-anchor outputs and ephemeral dust. With this change, commitment transactions don't pay any mining fee (in most cases), which gets rid of `update_fee` and all of the related channel reserve issues. It also gets rid of the undesired channel force-closes that happen when the mempool feerate spikes and channel participants disagree on what feerate to use, which has been a major source of wasted on-chain space.

It also offers better protection against pinning attacks (thanks to package relay) and reduces the on-chain footprint compared to anchor output channels.

We use a single anchor output whose amount is the sum of all trimmed outputs (outputs below the `dust_limit`, which are thus not included in the commitment transaction) and all millisatoshi amounts that have been rounded down to satoshis, capped at `240 sats` (and may thus be `0 sat` when there are no trimmed outputs or rounded down millisatoshi amounts). Note that when the sum of trimmed outputs and millisatoshi rounding exceeds `240 sats`, the commitment transaction won't by 0-fees anymore: but the fees aren't deduced from any of the commitment transaction's outputs.

This anchor output is unkeyed: it simply uses the standard P2A output introduced in https://github.com/bitcoin/bitcoin/pull/30352.

We remove the 1-block relative delay used by anchor output channels: this allows using our channel outputs (main balance or pending HTLCs) to CPFP a remote commitment transaction (no need to add external inputs, which simplifies wallet management and liquidity requirements).

Note that v3 transactions and pay-to-anchor outputs have been standard since the release of Bitcoin Core v28.0. Ephemeral dust has become standard in the Bitcoin Core v29.0 release. All of this is thus ready to use on the network now!

t-bast committed:
Zero-fee commitments using v3 transactions

We introduce a new `channel_type` that leverages v3 transactions,
pay-to-anchor outputs and ephemeral dust. With this change, commitment
transactions don't pay any mining fee, which gets rid of `update_fee`
and all of the related channel reserve issues. It also gets rid of the
undesired channel force-closes that happen when the mempool feerate
spikes and channel participants disagree on what feerate to use, which
has been a major source of wasted on-chain space.

It also offers better protection against pinning attacks and reduces the
on-chain footprint compared to anchor output channels. We use a single
anchor output whose amount is the sum of all trimmed outputs and rounded
down millisatoshi amounts (and may thus be `0 sat`), capped at 240 sats.
When the sum of trimmed outputs and rounded down millisatoshi amounts
exceeds 240 sats, the remaining amount directly goes to on-chain fees
(and in that case, the commitment transactions does pay some on-chain
fees on its own). We do this to ensure that only miners can collect
the excess from trimmed outputs, while respecting standardness rules.

We remove the 1-block relative delay used by anchor output channels,
because we don't need the CPFP carve-out rule when using v3 txs. This
allows using our channel outputs (main balance or pending HTLCs) to
CPFP a remote commitment transaction (no need to add external wallet
inputs, which greatly simplifies liquidity management).

v3 transactions and pay-to-anchor outputs have been standard since the
release of Bitcoin Core v28.0. Ephemeral dust has been standard since
the Bitcoin Core v29.0 release.

esneider commented on 05-onchain.md:
```diff
@@ -100,9 +100,11 @@ delayed_pubkey.
 4. _remote node's anchor output_: one output paying to the *remote node's*
 funding_pubkey.
-5. _local node's offered HTLCs_: Zero or more pending payments (*HTLCs*), to pay
+5. _shared ephemeral anchor_: one output paying to the *local* or *remote*
+node's funding_pubkey.
```
* The ephemeral in the output name isn't used anywhere else, probably just "shared anchor"?
* Description doesn't match the fact that the shared anchor is using P2A

morehouse replied:
This wasn't fully addressed.  It still refers to a keyed anchor instead of P2A.

saubyk commented:
> We use a single anchor output whose amount is the sum of all trimmed outputs (and may thus be 0 sat).

Hi @t-bast can you elaborate this a bit?

saubyk commented:
> > We use a single anchor output whose amount is the sum of all trimmed outputs (and may thus be 0 sat).
> 
> Hi @t-bast can you elaborate this a bit?

Never mind. This delving post has the details, makes for a good supplemental reading:
https://delvingbitcoin.org/t/which-ephemeral-anchor-script-should-lightning-use/1412

SomberNight commented on 03-transactions.md:
```diff
@@ -280,60 +298,73 @@ To redeem the HTLC, the HTLC-success transaction is used as detailed below. This
 * txin count: 1
    * `txin[0]` outpoint: `txid` of the commitment transaction and `output_index` of the matching HTLC output for the HTLC transaction
-   * `txin[0]` sequence: `0` (set to `1` for `option_anchors`)
+   * `txin[0]` sequence: `1` for `option_anchors`, `0xFFFFFFFD` for `zero_fee_commitments`, `0` otherwise
```
What is the difference between using `0` and `0xFFFFFFFD` here?
Neither imposes a bip68 timelock, and both explicitly opt-in to bip125 RBF.

I am just trying to understand the motivation for why the old-style txs were using `0` and why the newer style proposed here uses a different value.

SomberNight replied:
I see, so there is no interesting technical difference I missed. Thanks.
I understand the potential privacy aspect, but I don't think it's worth the bikeshedding in this case, so feel free to mark this as resolved :)

instagibbs replied:
I *suspect* anyone using TRUC tx will keep this 0 since it's not a meaningful signal, especially post-fullrbf world?

I don't think we have enough data to say exactly.

morehouse commented on 05-onchain.md:
```diff
@@ -148,9 +150,10 @@ A node:
-          - MUST spend any `to_local_anchor` output, providing sufficient fees as incentive to include the commitment transaction in a block.
-          Special care must be taken when spending to a third-party, because this re-introduces the vulnerability that was
-          addressed by adding the CSV delay to the non-anchor outputs.
+          - MUST spend any `to_local_anchor` or `shared_anchor` output, providing sufficient fees
+          as incentive to include the commitment transaction in a block. When using `nVersion=2`,
+          special care must be taken when spending to a third-party, because this re-introduces
+          the vulnerability that was addressed by adding the CSV delay to the non-anchor outputs.
```
Bringing up `nVersion=2` here is a bit confusing and may give the reader the idea they can set `nVersion=3` for `to_local_anchor` spends to avoid this issue (which won't relay).

```suggestion
          - MUST spend any `to_local_anchor` or `shared_anchor` output, providing sufficient fees
          as incentive to include the commitment transaction in a block. For `to_local_anchor`,
          special care must be taken when spending to a third-party, because this re-introduces
          the pinning vulnerability that was addressed by adding the CSV delay to the non-anchor outputs.
```

morehouse commented on 05-onchain.md:
```diff
@@ -515,7 +517,7 @@ A local node:
   - SHOULD extract the payment preimage from the transaction input witness, if
   it's not already known.
-  - if `option_anchors` applies:
+  - if `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions:
```
```suggestion
  - if `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions (i.e. `option_anchors` or `zero_fee_commitments` applies):
```

morehouse commented on 05-onchain.md:
```diff
@@ -532,8 +534,8 @@ A single transaction that resolves all the outputs will be under the
 
-Note: if `option_anchors` applies, the cheating node can pin spends of its
-HTLC-timeout/HTLC-success outputs thanks to SIGHASH_SINGLE malleability.
+Note: when `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions,
+the cheating node can pin spends of its HTLC-timeout/HTLC-success outputs.
```
```suggestion
Note: when `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` is used for HTLC transactions (i.e. `option_anchors` or `zero_fee_commitments`),
the cheating node can pin spends of its HTLC-timeout/HTLC-success outputs.
```

morehouse commented on 05-onchain.md:
```diff
@@ -588,25 +590,25 @@ Note: even if the `to_remote` output is not swept, the resulting
-If `option_anchors` does not apply to the commitment transaction, then
-HTLC-timeout and HTLC-success transactions are complete transactions with
-(hopefully!) reasonable fees and must be used directly.
+If `SIGHASH_ALL` is used for HTLC transactions, then HTLC-timeout and
+HTLC-success transactions are complete transactions with (hopefully!)
+reasonable fees and must be used directly.
```
This section describes *when* to use `SIGHASH_ALL` vs `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY`.  So we shouldn't assume the reader already knows which sighashes to use for each channel type.

```suggestion
If `option_anchors` and `zero_fee_commitments` do not apply, then HTLC-timeout and
HTLC-success transactions are signed with SIGHASH_ALL and are complete transactions with (hopefully!)
reasonable fees and must be used directly.
```

morehouse reviewed (commented):
Approach ACK.

Really eager to see this get implemented and used across the network.  Also hope we can follow up soon with the [changes needed](https://github.com/lightning/bolts/issues/1221#issuecomment-2626983064) to fix HTLC pinning.

TheBlueMatt reviewed (commented):
This was actually a good bit easier to implement than I thought.

Roasbeef commented:
> Just want to node that it was discussed in yesterday's spec call that as part of this work we'll probably also want to make option_simple_close transactions (optionally) v3, i.e., require v3 close transactions for v3 channels.

Compared to the HTLC case, are there any pinning concerns when it comes to coop close? The channel is no longer active, and the only outputs on the coop close transaction send funds directly to either party. 

I point this out as the TRUC constraints are passed on to all unconfirmed children. As a result, if the coop close transaction is v3, then a user wouldn't be able to CPFP with a non-v3 transaction to bump the fee (imagine the peer is just no longer offline). Another scenario that would be prevented is a non-v3 aware wallet spending that unconfirmed change to create a normal transaction, or fund a new channel.

carlaKC commented on 02-peer-protocol.md:
```diff
@@ -745,6 +744,7 @@ affect the channel operation).
 The currently defined basic types are:
   - `option_static_remotekey` (bit 12)
   - `option_anchors` and `option_static_remotekey` (bits 22 and 12)
+  - `zero_fee_commitments` (bit 40)
```
Why doesn't this channel type also have `option_static_remotekey`?

carlaKC replied:
> Why do you want to include the option_static_remotekey feature bit?

I thought that channel type is supposed to be all the features that a channel includes, so static remote key should be included here (because zero fee commitment channels will also use static remote keys).

It wasn't obvious to me that now that `option_static_remote_key` is assumed, it would also be assumed for channel types as well, but makes sense so that we can deprecate completely 👌

carlaKC replied:
> The issue with thinking that way is that we conflated channel types and lightning feature bits, so if we consider that 0-fee commitments use option_static_remotekey we end up being force to continue supporting the plain option_static_remotekey channel type, which we want to deprecate at some point (ideally, soon).

Yeah makes sense 👍  I was indeed conflating the two ideas.

carlaKC commented on 03-transactions.md:
```diff
@@ -171,9 +174,25 @@ After 16 blocks, anyone can sweep the anchor with witness:
+    OP_1 <0x4e73>
+
+The amount of this output is usually 0 sats, in which case it must be spent by the child transaction paying the mining fees.
+When the commitment contains [trimmed outputs](#trimmed-outputs), their amount is added to this `shared_anchor` output until it reaches 240 sats.
```
nit: I think it would be helpful to make a note here explaining why need to switch between the anchor and fees
* Ephemeral dust requires zero fees for values <= dust limit to be standard
* The dust limit for P2A is 240

Perhaps link to this issue: https://github.com/bitcoin/bitcoin/issues/31938 (helped me understand it better)

t-bast commented:
I have rebased to fix the merge conflicts, and squashed into a single commit. I have also:

- changed the explanation of the `240 sats` cap in Bolt 3
- added a warning in Bolt 5 about batching HTLCs transactions and the 10kvB limit

I'm working on adding a set of Bolt 3 test vectors.

carlaKC commented on bolt03/zero-fee-commitments-test.json:
```diff
@@ -0,0 +1,206 @@
+  "per_commitment_point": "0275d12130c276b4274358a328901f8fc47e6c72629102e4b46c9f27dd2c1dda98",
+  "remote_payment_basepoint_secret": "580bff39085f3a6ae8b1f32905e67366c522ea8f2418391145b2e98f1a7cb3f2",
+  "remote_htlc_basepoint_secret": "32df9c4dd46ab6210e74e81e15282106f8db883f45674eabb3324166c6513062",
+  "revocation_pubkey": "03e1ea1b8f5e4cae2dbac454a341ad5c3323b5827054b7e8b0497600010b31f860",
```
Could you provide the `revocation_basepoint` rather than the `revocation_pubkey` that's already been created from `revocation_basepoint` + `per_commitment_point`?

It's a bit hacky for me to override with the already derived pubkey in LDK.

t-bast replied:
That should be doable, I'll do that early next week.

t-bast replied:
That's great news! Do you have a branch of `ldk-server` that I can use for cross-compat tests?

tankyleo replied:
> That's great news! Do you have a branch of ldk-server that I can use for cross-compat tests?

So far have tested on mainnet with the PR in `ldk-node`, but will get a `ldk-server` branch ready.

carlaKC reviewed (approved):
> everything is working like a charm

🔥 

LGTM! Last test vector passes in LDK, just some nits and a request for a general "beware V3 gremlins" warning 🙏

t-bast commented on 02-peer-protocol.md:
```diff
@@ -2241,26 +2253,27 @@ is destined, is described in [BOLT #4](04-onion-routing.md).
-    its commitment transaction, it cannot pay the fee for the updated local or
-    remote transaction at the current `feerate_per_kw` while maintaining its
-    channel reserve.
+  - if `zero_fees_commitments` has not been negotiated:
```
Thanks, fixed in https://github.com/lightning/bolts/pull/1228/commits/a0bc7bc462927ebe6a01a5644208a36519a98ca6

tankyleo commented on bolt03/zero-fee-commitments-test.json:
```diff
@@ -197,6 +197,81 @@
+          "cltv_expiry": 920125
+        }
+      ],
+      "signed_commit_tx": "03000000000101ae1e3c841378cf9e4c383bfdd033d4b3c6945e0587ff16635a00b347eea2704b0100000000340fef800333000000000000000451024e7362841e0000000000160014f2123f1a4b67887f2e5f02eda73e6327010152eaea117a0000000000220020f2d298ffcfd6d899a3abada37bfc6f42ce0b7b66f3e39e903e8419ac97dca75a0400473044022033e3e272f752dd0f8eeb526f40b5c18dc44a528ba4a1257758baadee2a5e824b02207eeda638931de228f3 ...
```
52 sats on the anchor here too :)

t-bast commented on bolt03/zero-fee-commitments-test.json:
```diff
@@ -197,6 +197,81 @@
+          "cltv_expiry": 920125
+        }
+      ],
+      "signed_commit_tx": "03000000000101ae1e3c841378cf9e4c383bfdd033d4b3c6945e0587ff16635a00b347eea2704b0100000000340fef800332000000000000000451024e7362841e0000000000160014f2123f1a4b67887f2e5f02eda73e6327010152eaea117a0000000000220020f2d298ffcfd6d899a3abada37bfc6f42ce0b7b66f3e39e903e8419ac97dca75a04004830450221008bc280fdd03504d216662ed63e3c9e48fbe1d9e7be293ec01822f721646d6c4e022004ea9032d5f13113 ...
```
Hah, good thing you're reviewing this seriously, because the code I wrote was wrong, I didn't spend enough time on unit tests for millisatoshi rounding, and even the spec was wrong :facepalm:, thanks a lot for spotting this!

The anchor amount isn't just set based on trimmed outputs, but also on any millisatoshi amount that is rounded down to satoshis. I changed the spec in https://github.com/lightning/bolts/pull/1228/commits/fc83fef80763f5a7917ae320f41420a33e7dad96 to fix that, fix those 3 test vectors, and add another test vector for millisatoshi rounding without dust outputs.

Let me know if those test vectors work with LDK, and if I finally got this right this time!

tankyleo commented on 02-peer-protocol.md:
```diff
@@ -2241,26 +2253,27 @@ is destined, is described in [BOLT #4](04-onion-routing.md).
+      reserve (see [Updating Fees](#updating-fees-update_fee)).
+      - if `option_anchors` applies to this commitment transaction and the sending
+      node is the funder:
+        - MUST be able to additionally pay for `to_local_anchor` and 
```
nit: some trailing whitespace here

Roasbeef commented on 03-transactions.md:
```diff
@@ -281,60 +306,68 @@ To redeem the HTLC, the HTLC-success transaction is used as detailed below. This
+Before the commitment transaction outputs are determined:
+  - the base fee MUST be subtracted from the `to_local` or `to_remote` output,
+    as specified in [Fee Calculation](#fee-calculation).
+  - if `option_anchors` applies:
```
Should this actually be `zero_fee_commitments`? Don't understand this portion of the diff.

t-bast commented on 05-onchain.md:
```diff
@@ -537,8 +542,13 @@ A local node:
+    - If `zero_fee_commitments` applies:
+      - MUST NOT exceed the 10kvB limit of v3 transactions, otherwise the
+        transaction may not reach miners (as explained in the section
+        about [v3 limitations](#limitations-of-v3-transactions) below).
```
I'm not sure I understand your comment here. HTLC transactions are using v3, so if you batch a bunch of HTLC transactions together (using the fact that their input uses `SIGHASH_SINGLE | SIGHASH_ANYONECANPAY`), the resulting transaction is a batched HTLC transaction that is using v3. So regardless of whether the commitment transaction is confirmed or not, the batched HTLC transaction cannot exceed 10kvB. Does that make sense?

t-bast commented on 02-peer-protocol.md:
I can't see which section you're referring to here, github doesn't link that comment to any code change...

TheBlueMatt commented on 02-peer-protocol.md:
```diff
@@ -809,15 +809,19 @@ The sending node:
+    - set `dust_limit_satoshis` to a sufficient value to ensure that it is economical to spend, taking the cost of HTLC transactions into account.
+  - otherwise (not using `zero_fee_commitments` or `option_anchors`):
+    - set `feerate_per_kw` to at least the rate it estimates would cause the transaction to be immediately included in a block.
+    - set `dust_limit_satoshis` to a sufficient value to allow commitment transactions to propagate through the Bitcoin network.
```
this applies in both cases.

t-bast replied:
Indeed, moved back above where it applies to both cases in https://github.com/lightning/bolts/pull/1228/commits/e59e30ca504f9ff6999c80f66e0cddf132cd2197.

TheBlueMatt commented on 02-peer-protocol.md:
```diff
@@ -3253,6 +3270,9 @@ A sending node:
 A receiving node:
+  - if `zero_fee_commitments` is used:
+    - MUST ignore the message.
+    - SHOULD send a `warning` and disconnect.
```
I don't think this is useful tbh. Many nodes will send a commitment-update-message in a batch with the actual `commitment_signed`. As a result, on reconnect they have to resend the commitment update (they already signed it!) and we just enter a reconnect loop. We should instead fail the channel - the peer is hopelessly broken.

t-bast replied:
Makes sense, changed that to a channel fail in https://github.com/lightning/bolts/pull/1228/commits/e59e30ca504f9ff6999c80f66e0cddf132cd2197.

t-bast committed:
fixup! Zero-fee commitments using v3 transactions
